runtime/arm64.S

```diff
@@ -150,92 +154,178 @@ G(name):
+        stp     d4, d5, [TMP, 240]
+        stp     d6, d7, [TMP, 256]
+        stp     d16, d17, [TMP, 272]
+        stp     d18, d19, [TMP, 288]
+        stp     d20, d21, [TMP, 304]
+        stp     d22, d23, [TMP, 320]
+        stp     d24, d25, [TMP, 336]
+        stp     d26, d27, [TMP, 352]
+        stp     d28, d29, [TMP, 368]
+        stp     d30, d31, [TMP, 384]
+        add     TMP, TMP, #16
+        str     TMP, Caml_state(gc_regs)
+.endm
+
+/* Undo SAVE_ALL_REGS by loading the registers saved in Caml_state(gc_regs)
+   and refreshing ALLOC_PTR & TRAP_PTR from Caml_state */
+.macro RESTORE_ALL_REGS
+    /* Restore x0, x1, freeing up the next ptr slot */
+        ldr     TMP, Caml_state(gc_regs)
+        sub     TMP, TMP, #16
+    /* Restore registers */
+        ldp     x0, x1, [TMP, 16]
+        ldp     x2, x3, [TMP, 32]
+        ldp     x4, x5, [TMP, 48]
+        ldp     x6, x7, [TMP, 64]
+        ldp     x8, x9, [TMP, 80]
+        ldp     x10, x11, [TMP, 96]
+        ldp     x12, x13, [TMP, 112]
+        ldp     x14, x15, [TMP, 128]
+        ldp     x19, x20, [TMP, 144]
+        ldp     x21, x22, [TMP, 160]
+        ldp     x23, x24, [TMP, 176]
+        ldr     x25, [TMP, 192]
+        ldp     d0, d1, [TMP, 208]
+        ldp     d2, d3, [TMP, 224]
+        ldp     d4, d5, [TMP, 240]
+        ldp     d6, d7, [TMP, 256]
+        ldp     d16, d17, [TMP, 272]
+        ldp     d18, d19, [TMP, 288]
+        ldp     d20, d21, [TMP, 304]
+        ldp     d22, d23, [TMP, 320]
+        ldp     d24, d25, [TMP, 336]
+        ldp     d26, d27, [TMP, 352]
+        ldp     d28, d29, [TMP, 368]
+        ldp     d30, d31, [TMP, 384]
+    /* Put gc_regs struct back in bucket linked list */
+        ldr     TMP2, Caml_state(gc_regs_buckets)
+        str     TMP2, [TMP, 0]  /* next ptr */
+        str     TMP, Caml_state(gc_regs_buckets)
+    /* Reload new allocation pointer & exn handler */
+        ldr     ALLOC_PTR, Caml_state(young_ptr)
+        ldr     TRAP_PTR, Caml_state(exn_handler)
+.endm
+
 /* Allocation functions and GC interface */
         TEXT_SECTION(caml_system__code_begin)
         .globl  G(caml_system__code_begin)
 G(caml_system__code_begin):
 
+FUNCTION(caml_call_realloc_stack)
+        CFI_STARTPROC
+    /* Save return address and frame pointer */
+        CFI_OFFSET(29, -16)
+        CFI_OFFSET(30, -8)
+        stp     x29, x30, [sp, -16]!
+        CFI_ADJUST(16)
+        add     x29, sp, #0
+    /* Save all registers (including ALLOC_PTR & TRAP_PTR) */
+        SAVE_ALL_REGS
+        ldr     C_ARG_1, [sp, 16] /* argument */
+        SWITCH_OCAML_TO_C
+        bl      G(caml_try_realloc_stack)
+        SWITCH_C_TO_OCAML
+        cbz     x0, 1f
+        RESTORE_ALL_REGS
+    /* Free stack space and return to caller */
+        ldp     x29, x30, [sp], 16
+        ret
+1:      RESTORE_ALL_REGS
+    /* Free stack space and return to caller */
+        ldp     x29, x30, [sp], 16
+        add     sp, sp, 16 /* pop argument */
```

Comment:
I assume this SP adjustment is needed so that the backtrace is properly recorded.

Reply:
Yes, we want the stack to be well formed when we call `caml_raise_exn` in case a backtrace is needed.